src/sound_openal.cpp

```diff
@@ -567,16 +595,55 @@ class OpenALSoundManager: public ISoundManager
 		}
 		return playSoundRawAt(buf, loop, volume, pos);
 	}
+
 	void stopSound(int sound)
 	{
 		maintain();
 		deleteSound(sound);
 	}
+
+	void fadeSound(int soundid, float step, float gain)
+	{
+		m_sounds_fading[soundid] = FadeState(step, getSoundGain(soundid), gain);
+	}
+
+	void doFades(float dtime)
+	{
+		m_fade_delay += dtime;
+
+		if (m_fade_delay < 0.2f)
+			return;
+
+		float chkGain = 0;
+		for (UNORDERED_MAP<int, FadeState>::iterator i = m_sounds_fading.begin();
+				i != m_sounds_fading.end();) {
+			if (i->second.step < 0.f)
+				chkGain = -(i->second.current_gain);
+			else
+				chkGain = i->second.current_gain;
+
+			if (chkGain < i->second.target_gain) {
+				i->second.current_gain += (i->second.step * m_fade_delay);
+				i->second.current_gain = rangelim(i->second.current_gain, 0, 1);
+
+				updateSoundGain(i->first, i->second.current_gain);
+				++i;
+			} else {
+				if (i->second.target_gain <= 0.f)
+					stopSound(i->first);
+
+				m_sounds_fading.erase(i++);
```

Comment:
We're already using the same pattern in multiple places:
https://github.com/minetest/minetest/blob/master/src/serverenvironment.cpp#L1444
https://github.com/minetest/minetest/blob/master/src/nodedef.cpp#L1330
`std::map` only https://github.com/minetest/minetest/blob/master/src/particles.cpp#L448
And possibly a few more - haven't gone through all search results. Thanks for the review :)